File: 프로그래머스/1/72410. 신규 아이디 추천/신규 아이디 추천.py
def solution(new_id):
    answer = ''
    
    new_id = new_id.lower()     # 1단계
    new_id = list(new_id)
    
    # 2단계
    uniq = []
    for up_id in new_id:
        if not up_id.isalpha() and not up_id.isdigit() and up_id not in ['-', '_', '.']:
            uniq.append(up_id)            
    
    for u in uniq:
        new_id.remove(u)
    
    # 3단계
    if new_id:
        idx = 0
        while idx < len(new_id) - 1:
            if new_id[idx] == '.' and new_id[idx + 1] == '.':
                del new_id[idx]
            else:
                idx += 1
    
    # 4단계
    while new_id:
        if new_id[0] == '.':
            del new_id[0]            
        elif new_id[-1] == '.':
            del new_id[-1]
        else:
            break
    
    # 5단계
    if not new_id:
        new_id.append('a')
    
    # 6단계
    if len(new_id) >= 16:
        new_id = new_id[0:15]
    if new_id[-1] == '.':
        del new_id[-1]
    
    # 7단계
    if len(new_id) <= 2:
        while len(new_id) <= 2:
            new_id.append(new_id[-1])
            
    new_id = ''.join(new_id)
    
    return new_id

File: 프로그래머스/1/72410. 신규 아이디 추천/신규 아이디 추천.js
function solution(new_id) {
    var answer = ''
    answer = new_id.toLowerCase() // 1단계
    
        .replace(/[^a-z0-9-_.]/g, '') // 2단계
        /*
            [^...] - [^] : 대괄호 안에서 쓰면 제외의 뜻, ex) [^a-z] : 소문자 알파벳을 제외한 모든 문자
                   - 대괄호 안에서 -_. 같은 특수문자는 문자 그대로 해석     
            a-z : 소문자 알파벳 (a부터 z까지)
            0-9 : 숫자(0부터 9까지)
            g : 모든 검색 결과를 반환        
        */
    
        .replace(/\.+/g, '.') // 3단계
        /*
            \. = 문자 그대로 해석
            + = 바로 앞에 요소가 1번 이상 반복
        */
        .replace(/^\.|\.$/g, '') // 4단계
        /*
            ^ : 문자열의 시작
            $ : 문자열의 끝
            | : or 연산
        */
        .replace(/^$/, 'a') // 5단계
        .slice(0, 15).replace(/\.$/, ''); // 6단계
    
    while (answer.length <= 2) answer += answer[answer.length - 1]; // 7단계
    
    return answer
}